### PR TITLE
fix: Assignment of the Capacity Reservation id to an instance

### LINKED ISSUE #277: Creation of an Intance with a Capacity Reservation ID
## Description

When trying to call the module with the specification of a Capacity Reservation ID, the module ends up with the following error:

╷
│ Error: Invalid function argument
│ 
│   on .terraform/modules/terraform-aws-ec2-instance/main.tf line 45, in resource "aws_instance" "this":
│   45:           capacity_reservation_id = lookup(capacity_reservation_target.value, "capacity_reservation_id", null)
│     ├────────────────
│     │ capacity_reservation_target.value is "cr-xxxxxxxx"
│ 
│ Invalid value for "inputMap" parameter: lookup() requires a map as the
│ first argument.
╵

As per line 45 analysis, it looks like the bug is in the following statement:

capacity_reservation_id = lookup(capacity_reservation_target.value, "capacity_reservation_id", null)

it should rather be as follows:

capacity_reservation_id = lookup(capacity_reservation_target, "capacity_reservation_id", null)

In this case, terraform should lookup in the capacity reservation target map (since it is the iterator in the for_each) rather than in its value for the mapping of the attribute capacity reservation id

- [X] ✋ I have searched the open/closed issues and my issue is not listed.

## Versions

- Module version [Required]: 4.1.0

- Terraform version: v1.2.5

- Provider version(s): v4.22.0

## Reproduction Code [Required]

Call the module with following inputs:

module "ec2_capacity_reservation" {
  source = "../../"

  name = "${local.name}-capacity-reservation"

  ami                         = data.aws_ami.amazon_linux.id
  instance_type               = "u-6tb1.56xlarge"
  subnet_id                   = element(module.vpc.private_subnets, 0)
  vpc_security_group_ids      = [module.security_group.security_group_id]
  associate_public_ip_address = true

  capacity_reservation_specification = {
      capacity_reservation_target = {
        capacity_reservation_id = "cr-00000000000000000"
      }
    }

  tags = local.tags
}

 No
 Yes

## Expected behavior
TF plan should go through and the module should create an EC2 Instance which consumes the capacity reservation.

## Actual behavior

TF plan fails as it is unable to lookup for the capacity reservation id

### PR DESCRIPTION
**Pull request by schniber**
## Description
When trying to call the module with the specification of a Capacity Reservation ID, the module ends up with the following error:

╷
│ Error: Invalid function argument
│
│ on .terraform/modules/terraform-aws-ec2-instance/main.tf line 45, in resource "aws_instance" "this":
│ 45: capacity_reservation_id = lookup(capacity_reservation_target.value, "capacity_reservation_id", null)
│ ├────────────────
│ │ capacity_reservation_target.value is "cr-xxxxxxxx"
│
│ Invalid value for "inputMap" parameter: lookup() requires a map as the
│ first argument.
╵

As per line 45 analysis, it looks like the bug is in the following statement:

capacity_reservation_id = lookup(capacity_reservation_target, "capacity_reservation_id", null)

it should rather be as follows:

capacity_reservation_id = lookup(capacity_reservation_specification.value.capacity_reservation_target, "capacity_reservation_id", null)

In this case, terraform should lookup in the capacity reservation target map (since it is the iterator in the for_each) rather than in its value for the mapping of the attribute capacity reservation id

## Motivation and Context

Fixes #277 and #281 

## Breaking Changes

No breaking changes, just fixing a typo in line 45 in the main.tf

## How Has This Been Tested?
- [X] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [X] I have tested and validated these changes using one or more of the provided `examples/*` projects

- [X] I have executed `pre-commit run -a` on my pull request

### CODE REVIEW

**Comment by schniber** [comment]
@bryantbiggs : Hello Bryant, I noticed that my logic on the previous PR on the same topic was wrong.

In this PR I have revisited this statement in the main.tf:

`capacity_reservation_id = lookup(capacity_reservation_specification.value.capacity_reservation_target, "capacity_reservation_id", null)
`

I also tested thoroughly using the complete example and this is now satisfactory.

Sorry again for the double review work.

Enjoy your weekend !

**Comment by schniber** [comment]
Hello @bryantbiggs,

kind reminder for this PR which is ready for review.

Thanks again for your feedback.

Bests.

File: main.tf [hunk 1]
```diff
@@ -42,7 +42,7 @@ resource "aws_instance" "this" {
       dynamic "capacity_reservation_target" {
         for_each = lookup(capacity_reservation_specification.value, "capacity_reservation_target", [])
         content {
-          capacity_reservation_id = lookup(capacity_reservation_target, "capacity_reservation_id", null)
```
Comment (bryantbiggs) [inline main.tf:45]: I believe we should be able to copy this https://github.com/terraform-aws-modules/terraform-aws-autoscaling/blob/master/main.tf#L73-L86 to here

File: examples/complete/main.tf [hunk 2]
```diff
@@ -350,11 +350,11 @@ module "ec2_capacity_reservation" {
   instance_type               = "t3.micro"
   subnet_id                   = element(module.vpc.private_subnets, 0)
   vpc_security_group_ids      = [module.security_group.security_group_id]
-  associate_public_ip_address = true
+  associate_public_ip_address = false
 
   capacity_reservation_specification = {
     capacity_reservation_target = {
-      capacity_reservation_id = aws_ec2_capacity_reservation.targeted.id
```
Comment (bryantbiggs) [inline examples/complete/main.tf:357]: I think we want one targeted example https://github.com/terraform-aws-modules/terraform-aws-autoscaling/blob/master/examples/complete/main.tf#L468-L472

and one open example https://github.com/terraform-aws-modules/terraform-aws-autoscaling/blob/master/examples/complete/main.tf#L222-L224

so we'll need to add an open example

**Comment by schniber** [comment]
Hello @bryantbiggs 

Thanks for your valuable feedback that I have taken into account.

I have added the example for the targeted capacity reservation together with the simplification of the capacity reservation assignment. it's more readable with try function.

I have also deployed to test and well went through.

Looking forward for the merge if all is okay for you.

Thanks a ton !

Bests.

**Comment by schniber** [comment]
Hello @bryantbiggs 

Would it be possible to review the changes I made upon your feedback ?

Thanks again.

Bests.

**Review by bryantbiggs — APPROVED** [review]
Thank you!

**Comment by antonbabenko** [comment]
This PR is included in version 4.1.2 :tada:

**Comment by yewton** [comment]
Should the default value of `capacity_reservation_specification` be updated from `null` to `{}` ?

**Comment by bryantbiggs** [comment]
@yewton good spot - yes, looks like we need to update the variable as well. PRs are welcomed!

**Comment by schniber** [comment]
I will create a new PR to fix it.
Thanks for spotting it !

**Comment by yewton** [comment]
Thank you!

**Pull request merged**